src/fluka/fluka_funcs.cpp

```diff
@@ -0,0 +1,1104 @@
+      // if point and direction has changed, we should reset everything
+      reset_state(state);
+    } else {
+      // direction has changed reset ray history
+      state.history.reset();
+    }
+  }
+  /*
+  // direction changed reset history, may not be robust
+  if( dir[0] == state.old_direction[0] && dir[1] == state.old_direction[1] && dir[2] == state.old_direction[2] )  {
+    // position may have changed since last time, may not be robust
+    if( point[0] == state.old_position[0] && point[1] == state.old_position[1] && point[2] == state.old_position[2] ) {
+    } else {
+      // if point and direction has changed, we should reset everything
+      reset_state(state);
+    }
+  } else {
+    // direction has changed reset ray history
+    state.history.reset();
+  }
+  */
+
+  if (debug)
+    print_state(state);
+
+  if (state.on_boundary) {
+    // electrons do funny things on the boundary, since electrons can interact on the boundary, it may be the case
+    // that the when on the boundary the electron changes direction, not end up in the new region as originally thought
+    // when this happens
+    // Alfredo says that we only need to check the dot product of the normal of the last surface crossed &
+    // the particle direction if less than 0.0 then we return a step of 0.0 and set newRegion to -3
+
+    // since the fluka normal points the other way
+    if (state.prev_surface == 0 && state.next_surface != 0) {
+      if (debug)
+        std::cout << "Checking to see if we have changed direction using next surface" << std::endl;
+      if (dot_product(state.next_surface, point, dir) < 0.0) {
+        retStep = 0.0;
+        newRegion = -3;
+        return;
+      }
+    } else if (state.prev_surface != 0) {
+      if (debug)
+        std::cout << "Checking to see if we have changed direction using last surface" << std::endl;
+      if (dot_product(state.prev_surface, point, dir) < 0.0) {
+        retStep = 0.0;
+        newRegion = -3;
+        return;
+      }
+    } else {
+      fludag_abort("g_fire", "Failure from within FluDAG, next_surf & prev_surf are both 0", 1);
+    }
+
+    // if we are not on the boundary but think we should be
+    if (boundary_test(vol, point, dir) == 0) { // if ray not on boundary of leaving vol
+      if (debug)
+        std::cout << " ray is on boundary" << std::endl;
+      state.history.reset(); // reset history
+      state.on_boundary = false; // reset on boundary
+    }
+  }
+
+  // perform the actual ray fire
+  moab::ErrorCode rval = DAG->ray_fire(vol, point, dir, next_surf, next_surf_dist, &state.history); // fire a ray
+  if (rval != moab::MB_SUCCESS)
+    fludag_abort("g_fire", "Failure from within DAGMC, ray_fire returned errorcode", rval);
+
+  if (next_surf == 0) {   // if next_surface is 0 then we are lost
+    if (debug) {
+      std::cout << "!!! Lost Particle !!! " << std::endl;
```

Comment:
If this is an error, might be worth it to switch to using `std::cerr` rather than `std::cout`